stablehlo/dialect/TypeInference.cpp

```diff
@@ -97,6 +97,51 @@ bool tensorsHaveSameElType(Type type1, Type type2,
   return tensorsHaveSameElType({type1, type2}, ignoreFpPrecision);
 }
 
+unsigned potentiallyComplexBitWidth(Type type) {
+  auto complexTy = type.dyn_cast<ComplexType>();
+  return complexTy ? 2 * complexTy.getElementType().getIntOrFloatBitWidth()
+                   : type.getIntOrFloatBitWidth();
+}
+
+// Returns true if the element-type of type1 can be promoted to that of type2.
+// An element-type 'x' is promotatble to element-type 'y' is they have the same
+// base type and bitwidth(x) <= bitwidth(y). When 'x' and 'y' are quantized
+// element-types, then promotion is applied only to the 'storage_type'
+// component.
+bool isPromotableElementType(Type type1, Type type2,
+                             bool ignoreFpPrecision = false) {
+  auto tensorTy1 = type1.dyn_cast<ShapedType>();
+  auto tensorTy2 = type2.dyn_cast<ShapedType>();
+
+  if (!tensorTy1 || !tensorTy2) return false;
+
+  Type tensorEl1 = tensorTy1.getElementType();
+  Type tensorEl2 = tensorTy2.getElementType();
+
+  if (ignoreFpPrecision && tensorEl1.isa<FloatType>() &&
```

Comment:
What is ignore precision for?

Reply:
The ignore fl precision is previously used in `verifyReducerShape` (at https://github.com/openxla/stablehlo/blob/3bdd8593a46ababb362928187d58ccc418e15b99/stablehlo/dialect/TypeInference.cpp#L601 and https://github.com/openxla/stablehlo/blob/3bdd8593a46ababb362928187d58ccc418e15b99/stablehlo/dialect/TypeInference.cpp#L611) to match the behavior with HLO (https://github.com/openxla/xla/blob/8a0bc8775368467e37e853f4441a3be4c2610ae4/xla/service/shape_inference.cc#L148 and https://github.com/openxla/xla/blob/8a0bc8775368467e37e853f4441a3be4c2610ae4/xla/service/shape_inference.cc#L159).

The idea was to ignore the fp precision when fp types are involved. I plan was to keep the existing behavior for now and address its relevance as part of https://github.com/openxla/stablehlo/issues/369. Please let me know what do you think?